libvast/src/expression.cpp

```diff
@@ -206,9 +206,8 @@ caf::expected<expression> tailor(const expression& expr, const type& t) {
   auto x = caf::visit(type_resolver{t}, expr);
   if (!x)
     return x.error();
-  *x = caf::visit(type_pruner{t}, *x);
-  VAST_ASSERT(!caf::holds_alternative<caf::none_t>(*x));
-  return std::move(*x);
+  auto result = caf::visit(type_pruner{t}, *x);
+  return result;
```

Comment:
```suggestion
return caf::visit(type_pruner{t}, *x));
```

This could work.

Tailoring is a two-step process:

1. Resolve all type and key extractors
2. Remove all unneeded parts of the expression for the given type

It seems to me that (2) should be optional. If nothing can be pruned, then the pruning step should return an identity, i.e, the input expression. The output should `none` *iff* the input is *none*. To me, this seems to be the underlying bug.

Does that make sense?

Reply:
I'm not sure if I understand what "unneeded" means here: From looking at the code, I got the impression that we're trying to remove all parts of the expression that do not apply to the current type, because they would always return "false". (ie. in `foo.bar == "baz" || type.key == "asdf"`)

But then if no parts apply, this would mean that *nothing* is needed and the output should become `none`, no?

Reply:
> But then if no parts apply, this would mean that nothing is needed and the output should become none, no?

I think this is right, I had a different notion in mind. So the output *can* be `none` if everything gets "pruned away". Then the expression is effectively empty and will never match anything.

In historical queries, this in an "early abort". In continuous queries, it can't be. I think the reason for the assert comes from considering only the historical case.

We probably have to adjust the call site of `tailor` to do the right thing. But regardless, the assertion should be removed.